crates/environ/src/module.rs

```diff
@@ -133,8 +141,12 @@ impl TablePlan {
 // WARNING: when modifying, make sure that `hash_for_cache` is still valid!
 #[derive(Debug)]
 pub struct Module {
-    /// Unprocessed signatures exactly as provided by `declare_signature()`.
-    pub signatures: PrimaryMap<SignatureIndex, ir::Signature>,
+    /// This mapping lets us merge duplicate types (permitted by the wasm spec)
+    /// as they're declared.
+    pub signature_mapping: PrimaryMap<SignatureIndex, UniqueSignatureIndex>,
```

Comment:
This map :point_up: from signature indices into unique signatures (used :point_down: there), is the important change to `Module` worth noting. Note that I haven't added any methods here, which _might_ be a good idea, depending on what reviewers think. I tried to be conservative for now.